test/interactions/dragInteractionTests.ts

```diff
@@ -2,463 +2,457 @@
-
-        let target = component.background();
-        TestMethods.triggerFakeMouseEvent("mousedown", target, startPoint.x, startPoint.y);
-        TestMethods.triggerFakeMouseEvent("mousemove", target, endPoint.x, endPoint.y);
-        assert.isTrue(moveCallbackCalled, "callback was called on dragging (mousemove)");
-        assert.deepEqual(receivedStart, startPoint, "was passed the correct starting point");
-        assert.deepEqual(receivedEnd, endPoint, "was passed the correct current point");
-
-        receivedStart = null;
-        receivedEnd = null;
-        TestMethods.triggerFakeTouchEvent("touchstart", target, [{x: startPoint.x, y: startPoint.y}]);
-        TestMethods.triggerFakeTouchEvent("touchmove", target, [{x: endPoint.x, y: endPoint.y}]);
-        assert.isTrue(moveCallbackCalled, "callback was called on dragging (touchmove)");
-        assert.deepEqual(receivedStart, startPoint, "was passed the correct starting point");
-        assert.deepEqual(receivedEnd, endPoint, "was passed the correct current point");
-
-        svg.remove();
+      [TestMethods.InteractionMode.Mouse, TestMethods.InteractionMode.Touch].forEach((mode) => {
+        describe(`invoking with ${TestMethods.InteractionMode[mode]} events`, () => {
+          let callback: TestDragCallback;
+
+          beforeEach(() => {
+            callback = new TestDragCallback();
+            dragInteraction.onDragStart((start: Plottable.Point, end: Plottable.Point) => callback.call(start, end));
+          });
+
+          it("invokes onDragStart callback on start event", () => {
+            dragStart(startPoint, mode);
+            assert.isTrue(callback.getCalled(), "callback was called on beginning drag");
+            assert.deepEqual(callback.getStartPoint(), startPoint, "was passed the correct point");
+          });
+
+          it("does not invoke callback if drag starts outside Component", () => {
+            dragStart(positiveOutsidePoint, mode);
+            assert.isFalse(callback.getCalled(), "does not trigger callback if drag starts outside the Component (positive)");
+            dragStart(negativeOutsidePoint, mode);
+            assert.isFalse(callback.getCalled(), "does not trigger callback if drag starts outside the Component (negative)");
+          });
+
+          if (mode === TestMethods.InteractionMode.Mouse) {
+            it("does not invoke onDragStart on right click", () => {
+              TestMethods.triggerFakeMouseEvent("mousedown", component.background(), startPoint.x, startPoint.y, 2);
+              assert.isFalse(callback.getCalled(), "callback is not called on right-click");
+            });
+          }
+        });
       });
+    });
 
-      it("can de-register the onDrag() interaction", () => {
-        dragInteraction.attachTo(component);
-
-        let moveCallbackCalled = false;
-        let moveCallback = () => moveCallbackCalled = true;
+    describe("onDrag/offDrag", () => {
+      describe("registration", () => {
```

Comment:
do you think it's possible to refactor this a bit since `onDrag/offDrag`, `onDragStart/offDragStart`, and `onDragEnd/offDragEnd` all share "registration", "deregistration", "multiple registration", "deregistration isolation"